Callback handler for mlflow

mahmedk opened the pull request:
Description

Similar to the Wandb callback handler, this PR implements a handler for mlflow experiment tracking, logs all metrics and artifacts to mlflow server. Most of the util functions are reused from the wandb implementation.

- Optionally, creates a default `experiment` with name `langchain` if it doesnt already exist.
- Optionally, generates a random name and creates a run with it and tracks all metrics and artifacts against it.
- Records artifacts as `json` & `html` files.
- Supports specifying custom tracking server uri.

Example usage,
```
from langchain.callbacks import MlflowCallbackHandler
mlflow_callback = MlflowCallbackHandler(tracking_uri="sqlite:///mlruns.db")
manager = CallbackManager([StdOutCallbackHandler(), mlflow_callback])
llm = OpenAI(temperature=0, callback_manager=manager, verbose=True)
llm_result = llm.generate(["Tell me a joke", "Tell me a poem"] * 3)
mlflow_callback.flush_tracker(llm, finish=True)
```

mahmedk committed:
Callback handler for mlflow 

Implementation of callback handler for one of the most popular mlops platform - mlflow.
Inspired by wandb_callback handler.
This handler logs the metrics and artifacts to the mlflow server.

hwchase17 reviewed (commented):
this is awesome. an example notebook would be really helpful here

liangz1 reviewed (commented):
Hi @mahmedk, excited to see your efforts here! I'm from the MLflow maintainers team and we are actively working on adding new MLflow APIs to support recording text inputs, outputs, and LLM prompts, and it would be released with MLflow soon. With the new APIs, this PR could be largely simplified. I'm also working on the LangChain integration with MLflow and I'm happy to collaborate on this topic.

mahmedk commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,691 @@
+        self.action_records: list = []
+        self.nlp = spacy.load("en_core_web_sm")
+
+        self.metrics = {
```
> Hi @mahmedk, excited to see your efforts here! I'm from the MLflow maintainers team and we are actively working on adding new MLflow APIs to support recording text inputs, outputs, and LLM prompts, and it would be released with MLflow soon. With the new APIs, this PR could be largely simplified. I'm also working on the LangChain integration with MLflow and I'm happy to collaborate on this topic.

Hi @liangz1 , Thanks. Glad to have you looking at this PR. Super excited to hear that there will be native support for tracking prompts in MLFlow. I will be more than happy to collaborate.

mahmedk replied:
> Dumb question: Could you help me understand in which scenario would the user find the metrics helpful, and why would the user want to log these metrics?

These are some basic counters which records the actions taken by langchain for a text generating flow setup by user. User could use these for debugging, analysis ,  check the state of langchain resources like agents, chain & llms such as if they were called and if their start/end calls match.

akshaya-a commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,691 @@
+
+    def json(self, data: Any):
+        """To log all metrics in the input dict."""
+        for k, v in data.items():
```
can we convert this to a log_batch instead of repeated network requests in a loop?

mahmedk replied:
Done !

akshaya-a commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,691 @@
+                print("Could not save model.")
+                print(repr(e))
+                pass
+            self.mlflg.artifact(langchain_asset_path)
```
I'm not sure if you're planning on coordinating merging this PR with @liangz1 after the flavor is implemented (seems like that would be the easiest), but if not you can implement a flavor in "userspace" something like this: https://github.com/akshaya-a/langchain/pull/1/files#diff-a8185aaebbb28c2ebd37436cf39390e57ede177bc301cb7dc897b90f11d3b2bbR134

liangz1 replied:
Hi @akshaya-a , I tried out your feature branch, it works well with some minor fixes and the logged data looks awesome. Would you like to file a PR for it? Besides, it would be great if you can use the langchain flavor released with mlflow 2.3 to log the model.

akshaya-a replied:
hey @liangz1 , no I think @mahmedk's PR is similar enough and he's responsive with feedback so I'm planning on just abandoning my PR. Hope that's ok with you both! Also FYI I'm AK on the AzureML team working with Jin + your team so we're in sync :)

@mahmedk let me know if you want me to contribute any specific things to your PR (the maintainer seems to be mainly awaiting the sample notebook like wandb) to parallelize efforts.

mahmedk replied:
@akshaya-a thanks ! Yes, am pending to complete the notebook. Haven't been able to spend time on it. If you could do that or if you have already done it, please point me to it, I will use it and test the PR.

akshaya-a commented:
super excited for this @mahmedk, great work!

mahmedk committed:
Report metrics in batch

Update code to report metrics in batch
Support to include 'step' in the metric

mahmedk commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,691 @@
+            return
+        self.mlfc.log_metric(self.run.info.run_id, key, value)
+
+    def json(self, data: Any):
```
Done !

liangz1 commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,692 @@
+        else:
+            raise ValueError("Unexpected data format provided!")
+
+    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
```
Shall we log the model using the mlflow langchain flavor? https://github.com/mlflow/mlflow/blob/master/mlflow/langchain/__init__.py

mahmedk replied:
yes @liangz1

mahmedk replied:
@liangz1 added support for mlflow langchain flavor.

mahmedk committed:
Support for mlflow langchain flavor

- Added support to report chain assets using mlflow.langchain.log_model. Using log_artifact for other types.
- Fixed 'input' key error when prompt templates are used.
- Improved the reporting logic by using mlflow active run context

mahmedk commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,692 @@
+
+        self.mlflg.metrics(self.metrics, step=self.metrics["step"])
+
+        chain_input = inputs["input"]
```
@liangz1 please try with latest changes.
fixed the ``` KeyError: 'input' ``` issue.
For ``` mlflow.exceptions.MissingConfigException: Yaml file './mlruns/1/meta.yaml' does not exist. ``` . Please delete if the db and mlruns folder already exist and then try.

akshaya-a commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,719 @@
+        """Initialize callback handler."""
+        import_pandas()
+        import_textstat()
+        spacy = import_spacy()
```
some of the metrics computed here are rather heavy which is why I also looked at hwchase's approach to hide the spacy import/computations behind the vizualize/complexity booleans that can be turned on/off. Consider a similar approach to separate "core tracking" capabilities from "evaluation" which might require more heavy dependencies/computation

https://github.com/akshaya-a/langchain/pull/1/files#diff-a8185aaebbb28c2ebd37436cf39390e57ede177bc301cb7dc897b90f11d3b2bbR199

mahmedk replied:
Agree, Will do the required change.

akshaya-a commented on langchain/callbacks/mlflow_callback.py:
```diff
@@ -0,0 +1,719 @@
+        resp.update(flatten_dict(serialized))
+        resp.update(self.metrics)
+
+        self.mlflg.metrics(self.metrics, step=self.metrics["step"])
```
minor point - this will also log a metric called "step" i think, so you might want to drop it from the dict before logging

mahmedk replied:
yeah true. Will pop out that metric. Thanks !

mahmedk commented:
Closing in favor of https://github.com/hwchase17/langchain/pull/4150
